src/bind.py

```diff
@@ -196,37 +239,89 @@ def handle_relation_data(
         Returns:
             A resulting DNSRecordProviderData to put in the relation databag
         """
-        # Create staging area
-        tempdir = tempfile.mkdtemp()
-
-        # Write zone files
-        zones = {}
+        # Create zones list
+        zones: typing.List[Zone] = []
         for record_requirer_data, _ in relation_data:
-            zones.update(self._to_bind_zones(record_requirer_data))
-        logger.debug("ZONES: %s", zones)
-        for name, content in zones.items():
-            self._write(pathlib.Path(tempdir, f"db.{name}"), content)
-
-        # Write the named.conf file
-        self._write(
-            pathlib.Path(tempdir, "named.conf.local"), self._generate_named_conf_local(list(zones))
-        )
+            zones.extend(self._record_requirer_data_to_zones(record_requirer_data))
+
+        # Check for conflicts
+        _, conflicting = self._get_conflicts(zones)
+        if len(conflicting) > 0:
+            return self._create_dns_record_provider_data(relation_data)
 
-        # Move the valid staging area files to the config dir
-        for file_name in os.listdir(tempdir):
-            shutil.move(
-                pathlib.Path(tempdir, file_name), pathlib.Path(constants.DNS_CONFIG_DIR, file_name)
+        # Create staging area
+        with tempfile.TemporaryDirectory() as tempdir:
+            # Write zone files
+            zone_files: typing.Dict[str, str] = self._zones_to_files_content(zones)
+            for name, content in zone_files.items():
+                pathlib.Path(tempdir, f"db.{name}").write_text(content, encoding="utf-8")
+
+            # Write the named.conf file
+            pathlib.Path(tempdir, "named.conf.local").write_text(
+                self._generate_named_conf_local([z.domain for z in zones]), encoding="utf-8"
             )
 
+            # Move the valid staging area files to the config dir
+            for file_name in os.listdir(tempdir):
+                shutil.move(
+                    pathlib.Path(tempdir, file_name),
+                    pathlib.Path(constants.DNS_CONFIG_DIR, file_name),
+                )
+
         # Reload charmed-bind config
         self.reload()
```

Comment:
Is there a validation that could be done before the reload or a check if the reload failed?

Reply:
I'll add more validation in subsequent PRs. For now the charm will end up in error state. That said, if the implementation is correct (which should be validated by unit tests), the output will be only zonefiles and the DNS configuration will therefore be valid. So if it fails, I'm thinking that it's right that it goes into an error state as it will be not recoverable